### ISSUE TITLE
Proposal: Support editing content viewed by user using this extension

### ISSUE THREAD

**Issue by santhoshtr** [wikimedia/mediawiki-extensions-MobileFrontendContentProvider#1]
The section translaiton team recently had a requirement of having the production wiki articles in local development environment just before we publish a section. All our APIs are pointing to production wikis. So our local development environment will behave like we are doing the translation in a production wiki. Except at publishing stage. When we publish a section, it can only be at the local wiki and it should be on top of an existing article(not a new article creation). For this we did something similar to what this extension does.

We fetch the content of production wiki using a hook just before publishing. https://gerrit.wikimedia.org/r/c/mediawiki/extensions/ContentTranslation/+/701894

So I was thinking whether this extension can be enhaned further to give such reading+editing simulation in test or development environments. May be by incorportating the above linked code to MobileFrontendContentProvider extension.  What do you think?

BTW, is there any reason this extension to be MobileFrontend dependent(in naming), I see that as a general content provider for wikis.

**Comment by jdlrobson** [comment]
Definitely, something I want to help with as this code was envisioned to support as many extensions as possible. Right now this code (well the code currently in MobileFrontend that I'm trying to pull out) is being used on https://patchdemo.wmflabs.org/. Would be great to see ContentTranslation in that tool and for this extension to help proxy content there too.  If you have a PHP function returning a string of HTML representing the article via OutputPage, then this would be the extension for that. I'd just need to understand your code a bit more to help make that work.

For JavaScript API-driven experiences, we've also been using  configurable APIs e.g. [Nearby](https://github.com/wikimedia/mediawiki-extensions-NearbyPages/blob/master/extension.json#L24), [Popups](https://github.com/wikimedia/mediawiki-extensions-Popups/blob/master/extension.json#L72), [Vector.](https://github.com/wikimedia/Vector/blob/master/skin.json#L236) This works on the mobile editor for example, but only in a read-only capacity. For saving, we haven't had that use case, but I'd suggest having a separate config variable for saving. Would that kind of approach work for you?


> BTW, is there any reason this extension to be MobileFrontend dependent(in naming), I see that as a general content provider for wikis.

Mostly historic reasons.

We originally built this out for mobile, but expanded it to cover desktop as part of desktop improvements. e.g. https://patchdemo.wmflabs.org/wikis/d09e6411b5/wiki/Spain 

This could be made to be MobileFrontend dependent, by adding an equivalent hook (and IContentProvider class to core) but it's essential it needs to run before MobileFrontend's OutputPageBeforeHTML hook does. More use cases and making the needs more generic would definitely help with that.
